massa-protocol-worker/src/protocol_network.rs

```diff
@@ -247,23 +247,25 @@ impl ProtocolWorker {
             // Add the ops of info.
             info.operations = Some(operation_ids.clone());
 
-            let missing_operations = operation_ids
-                .into_iter()
-                .filter(|op| {
-                    // Filter missing operations returning true if not found and compute
-                    // the sum of operation bytes size
-                    if let Some(operation_id) = self.checked_operations.get(&op.into_prefix()) {
-                        let op = self
-                            .storage
-                            .retrieve_operation(operation_id)
-                            .expect("unexpected checked operation not in storage");
-                        info.operations_size =
-                            info.operations_size.saturating_add(op.serialized_size());
-                        return false;
-                    }
-                    true
-                })
-                .collect();
+            let missing_operations = {
+                let ops = self.storage.read_operations();
```

Comment:
I prefer the previous `retrieve_operation(operation_id)` API of storage, because here you are holding a (read)lock over all operations in storage while looping over the ids and processing each one, which could result in loss of parallelism.

On the other hand, acquiring a lock for an individual operation at each iteration should be ok, especially since `parking_lot` is optimized for the uncontended case.

> Uncontended lock acquisition and release is done through fast inline paths which only require a single atomic operation (source: https://docs.rs/crate/parking_lot/latest).

This comment sort of applies also to the entire structure of indexes inside storage, since previously we had map containing other arcs/locks, which allowed for quick retrieval of a reference, followed by th acquisition of a lock around a single object. 

Now we have a single lock around for example a `BlockIndexes`, and the coarser lock makes it harder to achieve parallelism in principle. 

Parking lot is designed around the idea of have a lot of very fine grain locks, whereas it now seems we are rather going for a coarser grained(and more complex because of the manual reference counting) design, and I'm not sure why actually. Maybe you can explain the rationale behind this?

Reply:
Thanks for the comments !

One problem with `retrieve_operation` is that things might have changed within a loop (ie the block containing the list of operations we were reading was dropped). Now we can achieve what we used to do with retrieve_operation easily, but we can also keep a finer locking for this kind of situations.

As for the performance impact, it's a real question, and is hard to know a priori. Here is a little playground to test the perf impact of parking_lot mutexes: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2b7c64dfaab0c5d4aefc006a1dab83b4
Looks like we get a x4 performance gain by keeping the same lock.
Even if it uses a single atomic operation, it is a big CPU-wide sync impact, and a common cause of cache misses.

Regarding  per-element locks, we used to have them only around blocks (not ops/endorsements) because we wanted blocks to be subject to interior changes, but this is not the case anymore as the only thing we do is add/remove elements from the global indexes so we need this global lock on the container. Note that we use RwLocks to make the very frequent concurrent reads more efficient.

Reply:
@gterzian  interesting read: https://spcl.inf.ethz.ch/Publications/.pdf/atomic-bench.pdf

Main take-away: 

`Atomics prevent any instruction level parallelism, significantly limiting the bandwidth (up to 30x in comparison with simple writes), even if there are no dependencies between the successive operation`

Reply:
> Here is a little playground to test the perf impact of parking_lot mutexes

Yes taking the lock, and then looping, is going to be faster than taking the lock at each iteration, from the perspective of a single thread. 

However from the perspective of a multi-threaded system, the one "fast" thread could be causing every other thread to wait for the lock for the duration of the loop. 

So it's not about performance of a single thread, it's about opportunity for parallelism. By the way, if the system is designed around opportunities for parallelism, it is usually quite easy to optimize parts, like some kind of hot loop. However, if the system is not designed for parallelism, it is very hard to introduce it later...

> Basically, any atomic operation might temporarily impact all cores, and might cause cache misses

Interesting article, although I've just scanned the conclusion. 

It seems to apply mostly to parallel data structures, like a parallel map, that do a lot of atomic operations, and less-so to simply having a lock do a single operation(in the fast uncontended case as is the case with parking lot), followed by simply using the underlying locked structure without further atomic operations. 

Basically, it's a good argument against using complicated parallel data structures, vs simple locks. Thanks for sharing it. 

@damip